### ISSUE TITLE
Handle ARGV "correctly"

### ISSUE THREAD

**Issue by Pookah-Productions** [nushell/nushell#10781]
### Related problem

I decided try Nushell for a script I'm writing that will help me with a large, complex project I'm working on. The script, among other tasks, uses templates to generate new files in the appropriate directories, run various commands in sequence and the like. It's a mix of Zsh and Typescript. It has a fairly complex CLI, with lots of commands, à la Git or Rails, along with some accompanying flags. Those commands have a veritable smorgasbord of subcommands and flags. I tried using the `commandline` command, but couldn't figure out how to make that work. I also tried the suggested solution, using `main[ ...args ]`. That works as long as I don't use any flags. But `main [ ...args, ...flags ] doesn't work, either! But adding the flags as discrete elements to the parameter list for `main` is, well, arduous and error prone. Not to mention long! Very long.

A fairly thorough search of the documentation and the issue queue showed some requests for similar functionality, but I haven't come across any viable solution.

### Describe the solution you'd like

I'd love to use ARGV as it is used in other script languages.

For instance, in Typescript:

```ts
const argv = process.argv
```

I can use a custom script to parse `argv` and do the work that needs to be done.

### Describe alternatives you've considered

Basically, if there isn't a solution, I have to go back to using my complicated mix of Zsh and Typescript! I'd really like to not have to do that! 😵‍💫

**Comment by fdncred** [comment]
`commandline` is written to modify the repl input buffer.

If you search our repo for `commandline --` you'll see a bunch of tests with it.

Also, here's some keybindings I use that utilize `commandline`
```
  {
    name: insert_last_arg_from_prev_cmd
    modifier: alt
    keycode: char_.
    mode: [emacs, vi_normal, vi_insert]
    event: {
        send: executeHostCommand
        cmd: "commandline --insert (history | last | get command | parse --regex '(?P<arg>[^ ]+)$' | get arg | first)"
    }
  }
  {
    name: fuzzy_history_fzf
    modifier: shift_alt
    keycode: char_r
    mode: [emacs , vi_normal, vi_insert]
    event: {
      send: executehostcommand
      cmd: "commandline (
        history 
          | each { |it| $it.command } 
          | uniq 
          | reverse 
          | str join (char -i 0) 
          | fzf --read0 --tiebreak=chunk --layout=reverse  --multi --preview='echo {..}' --preview-window='bottom:3:wrap' --bind alt-up:preview-up,alt-down:preview-down --height=70% -q (commandline) 
          | decode utf-8 
          | str trim
      )"
    }
  }
  {
    name: insert_file
    modifier: alt
    keycode: char_y
    mode: [emacs vi_normal vi_insert]
    event: {
      send: executehostcommand
      cmd: "commandline --insert (fzf --tiebreak=chunk --layout=reverse  --multi --preview='echo {..}' --preview-window='bottom:3:wrap' --height=70% | decode utf-8 | str trim)"
    }
  }
  {
    name: insert_sudo
    modifier: alt
    keycode: char_s
    mode: [emacs, vi_insert, vi_normal]
    event: [
      { edit: MoveToStart }
      { send: ExecuteHostCommand,
        cmd: 'if (commandline | split row -r '\s+' | first) != `sudo` { commandline --insert `sudo `;commandline --cursor 100 }'
      }
    ]
  }
```

I'm not sure `commandline` is what you want though. It doesn't sound like you want to modify the repl buffer.

I don't believe we really allow you to parse your own parameters. You have to use `...rest` and `--long(-l)` flags in custom commands. Otherwise, nushell wouldn't be able to do the type checking that it does. I think this is pretty exhaustive and up to date https://www.nushell.sh/book/custom_commands.html#custom-commands

**Comment by Pookah-Productions** [comment]
Yeah, it suddenly dawned on me, in the middle of the night!, that `commandline` was the wrong thing. I totally missed the examples, sorry. I guess my looking through them wasn't quite as thorough as it could've been!

I have to admit to being somewhat surprised that a scripting language doesn't allow for the parsing of parameters. I was beginning to suspect that the opaqueness was a deliberate decision. A small note on the custom-commands page telling prospective users that a "raw" ARGV is unavailable would help others, especially as CLI apps become more popular and complex. I did spend a fair amount of time playing around with Nushell, and liked what I saw, which is why I gave it a go. Despite the considerable number of clues, I failed to pick up on the fact that you can't get a "raw" ARGV! A definite "Couldn't see the forest for the trees", kind of a thing, I guess. And while that's on me, a note in the docs saying such and such would've stopped me in my tracks. To be honest, I probably wouldn't have even given Nushell a go. 

Just as a suggestion, the Nushell team might want to reconsider their decision on ARGV? I was hoping to use Nushell for its performance, the very interesting philosophy behind it and the fact that it's way more contemporary than every other scripting language out there. Scripting needs a new approach; we're stuck with 1980's ideas, a syntax that's more cuneiform than explanatory and complex rules about, well, just about everything! Nushell could be a way forward. But having to adjust parameters in a `main [...]` function to define a CLI, and I've no idea how multiple subcommands could be done, is a definite barrier.

Thanks for your time and reply, @fdncred.

**Comment by fdncred** [comment]
Thanks for your feedback. We're making this up as we go so just because we don't have an ARGV thing now doesn't mean we won't have one in the future. Maybe @kubouch would have different suggestions and he would know better than I to know about the nushell viability of having ARGV.

**Comment by kubouch** [comment]
You can define `def --wrapped main [...rest] { ... }`. Unknown flags will be collected as strings into `$rest`.

**Comment by fdncred** [comment]
oh ya, nice! i forgot about that. seems like that works like expected, at least my understanding of it.
```nushell
❯ def --wrapped main [...rest] { $rest }
❯ main a b c -d 5 -e -f -g --h 08:38:59 AM
╭─┬───╮
│0│a  │
│1│b  │
│2│c  │
│3│-d │
│4│  5│
│5│-e │
│6│-f │
│7│-g │
│8│--h│
╰─┴───╯
```

**Comment by kubouch** [comment]
Subcommands are probably best done via nested submodules see https://www.nushell.sh/book/modules.html#main.

If you find our documentation lacking somewhere, feel free to submit a PR.

**Comment by Pookah-Productions** [comment]
Thanks, @kubouch! That worked. 

I'd actually decided to simply "front" the .nu script with a bash script that replaced all the spaces in $* with a percent sign, and pass that as a string. That... was a work around, but your solution is so much better!

I now have this...

```
# The main entry point.
def --wrapped main [...argv] {
  $argv
}
```

That gives me: 

```
╭───┬───────────╮
│ 0 │ generate  │
│ 1 │ -xyz      │
│ 2 │ --verbose │
╰───┴───────────╯
```


Thanks, @fdncred, for your help!

I will think about a page for your documentation. If my day and weekend pans out as it should, I hope to have something for you by Sunday evening, Eastern Time.

Thanks, again!

**Comment by kubouch** [comment]
OK, cool. Just note that the book should be as compact as possible. Rather than adding swaths of text, it is preferable to try refactoring the current text to have more useful information while maintaining succintness.

This could easily be a note in “Custom Commands”, for example, not necessary to have a page for it. Now looking at it, we don't even have `main` usage in scripts documented, only in modules (you can use module as a script!). This could also be improved.

**Comment by Pookah-Productions** [comment]
Thanks, I'll take a look at that!
